src/frontend/src/optimizer/plan_node/mod.rs

```diff
@@ -419,29 +419,31 @@ impl PlanTreeNode for PlanRef {
     }
 }
 
-impl StreamPlanRef for PlanRef {
-    fn distribution(&self) -> &Distribution {
-        &self.plan_base().dist
+impl PlanNodeMeta for PlanRef {
+    fn node_type(&self) -> PlanNodeType {
+        self.0.node_type()
     }
 
-    fn append_only(&self) -> bool {
-        self.plan_base().append_only
+    fn plan_base(&self) -> &PlanBase {
+        self.0.plan_base()
     }
 
-    fn emit_on_window_close(&self) -> bool {
-        self.plan_base().emit_on_window_close
+    fn convention(&self) -> Convention {
+        self.0.convention()
     }
 }
 
-impl BatchPlanRef for PlanRef {
-    fn order(&self) -> &Order {
-        &self.plan_base().order
+/// Implement for every type that provides [`PlanBase`] through [`PlanNodeMeta`].
+impl<P> GenericPlanRef for P
+where
+    P: PlanNodeMeta + Eq + Hash,
+{
```

Comment:
How about defining a new trait to abstract `PlanNodeMeta + Eq + Hash`?

Reply:
These codes will be soon cleaned up in my next PRs: #13000. 🥵 It may be acceptable to leave it as it is now.